### ISSUE TITLE
Sentence throughput regressions

### ISSUE THREAD

**Issue by graemenail** [hplt-project/OpusTrainer#33]
Current we don't have any measures on the performance impact of code changes. 

We should have a test that benchmarks e.g. run-time to process 1M sentences for each filter. Filters would need to provide a representative (set of) configurations to test over.

**Comment by XapaJIaMnu** [comment]
We should record typical marian training speed (WPS) on a given hardware and compare what is the throughput we can achieve with OpusTrainer. We are probably quite suboptimal with heavy use of placeholders.

**Comment by XapaJIaMnu** [comment]
This is critical,  we are wasting tremendous amount of GPU time because we wait on data production.  GPUs idle when training student models, essentially

**Comment by graemenail** [comment]
We should put some numbers here. How does it scale with (opustrainer) batch size?

**Comment by XapaJIaMnu** [comment]
I'll do some profiling. This didn't use to be this slow when doing terminology, i'll have to see what happened.

**Comment by XapaJIaMnu** [comment]
Some numbers:
```yml
# Datasets are already TSV files
datasets:
  clean: /tmp/clean.zhen.100krep

stages:
  - start

start:
  - clean  1
  - until clean 1

#modifiers:
  #- UpperCase: 0.1
  #- TitleCase: 0.1
  #- Tags: 0.1
  #  custom_detok_src: zh

seed: 1111
trainer: cat
```
Varying application of tags (with and without detokenization), UpperCase and TittleCase (albeit those would only be applied on the target side but oh well good enough for benchmarking)

command used:
```bash
time opustrainer-train -c CONFIG.YML -d cat > /dev/null
```
Checked current main and a version from 1.5 months ago when the logger was merged:

| Type             | 1f287a5 | f695e04 |
|------------------|--------:|--------:|
| No Modifier      |      6s |      6s |
| UpperCase        |      6s |      6s |
| TitleCase        |      8s |      8s |
| Tags             |    205s |     27s |
| Tags + detok_src |    287s |    121s |

We have 9x regression of performance when using tags after the tags rewrite, plus the detokenizer is really slow. We probably want the detokenizer to be multithreaded.

**Comment by graemenail** [comment]
diff: https://github.com/hplt-project/OpusTrainer/compare/f695e04586c3875d24c8edd1db2b37e57fb4af1f...1f287a5f73fc8d994fedd55c00dfe823ee5ac085

Thanks for generating these. Hopefully with the diff we can pin down what is causing the slowdown.

**Comment by jelmervdl** [comment]
It's the retokenisation, mostly: https://share.firefox.dev/3ExGcxB

**Comment by jelmervdl** [comment]
I think the more logical design would probably be to move this loop:

```py
for modifier in modifiers:
    batch = list(try_trace_map(lambda line: modifier(line.rstrip('\r\n')) + '\n', batch))
```

Into a multiprocessing pool. Probably ideally by splitting the batch across multiple workers, each worker moving a part of the batch through all modifiers.

And the difficult bit is to do this in such a way that even when the pool expands, or the order of how lines are split across the workers changes, the random seed stays consistent per line. Otherwise I don't know how we achieve reproducibility.

In this model only lines need to be pickleable I think, which is easy because they're just strings. Modifiers are loaded before the fork() I hope, and thus should be safe.

**Comment by XapaJIaMnu** [comment]
This is exactly what I did quickly but it claimed various things needed to be pickleable and I was planning to go over them one by one.

**Comment by jelmervdl** [comment]
Hahha, I have run into an interesting problem now:

The loop, with workers, is essentially:
```py
for seed, sub_batch in group(batch):
  random.seed(seed)
  for modifier in modifiers:
      for line in batch[sub_batch]:
        batch[line] = modifier(line)
```

If `group(batch)` generates more or fewer sub-batches, or the workers pick up the batches in different orders, the order in which `random.random()` is called inside the `modifier(line)` call is different.

So, I think I need to pre-generate a seed for each line. Problem is then that I also have to flip the loop, e.g.:
```py
for sub_batch in group(batch):
  for seed, line in sub_batch:
    random.seed(seed)
    for modifier in modifiers:
      batch[line] = modifier(line)
```

**Comment by XapaJIaMnu** [comment]
i mean, full determinism is nice to have, but realistically we probably don't care about it that much.

We should have some deterministic mode just so that we can pass tests reliably/discover breakages but it's not extremely important.

**Comment by jelmervdl** [comment]
I think I'll rewrite it a bit to do batch-based seeding, and modify the modifiers to take in batches. A change that was necessary anyway, and opens up the room for many more optimisations.

In terms of deterministic, I'm settling for deterministic unless you start changing parameters like any of the modifier parameters (obvious) or batch sizes (less obvious, but understandable when you think about it). I'll try to keep it deterministic between runs with exactly the same config/arguments, and ideally when you mess around with the number of workers. I think that's achievable.

**Comment by jelmervdl** [comment]
I'll make another pull request trying to integrate https://github.com/marketplace/actions/continuous-benchmark or something so we won't end up in this situation again :D
